### PR TITLE
fix: password validation

### LINKED ISSUE #9215: [CAL-1826] App Signup page wrong password validation
* [Wrongly tells that one character password is acceptable by showing the hint in green](https://www.loom.com/share/1bf5c9af3e294092b2098079a9326e29)
* It looks like there is just number of characters as validation. I was able to just set 1234567 as password

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-1826](https://linear.app/calcom/issue/CAL-1826/app-signup-page-wrong-password-validation)</sub>

### PR DESCRIPTION
**Pull request by piyushgarg-dev**
## What does this PR do?
Added Password Validation in Signup

Fixes #9215

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How should this be tested?
- Go to https://app.cal.com/signup
- Try invalid password and you would see errors
- Correct password would let you signup

### CODE REVIEW

**Review by sean-brydon — CHANGES_REQUESTED** [review]
Great effort! lets use zod to keep consistent with the rest of our forms 

https://github.com/react-hook-form/resolvers#Zod

File: apps/web/pages/404.tsx [hunk 1]
```diff
@@ -291,7 +291,7 @@ export default function Custom404() {
                       className="relative flex items-start space-x-4 py-6 rtl:space-x-reverse">
                       <div className="flex-shrink-0">
                         <span className="bg-muted flex h-12 w-12 items-center justify-center rounded-lg">
-                          <Slack strokeWidth="1.5" fill="currentColor" className="h-6 w-6" />
+                          <Slack strokeWidth="1.5" fill="currentColor" className="h-6 w-6 dark:text-white" />
```
Comment (Udit-takkar) [inline apps/web/pages/404.tsx:294]: I think you made this change in one of your earlier PR. This is not required here

**Review by Udit-takkar — COMMENTED** [review]
@piyushgarg-dev remove yarn.lock file

File: apps/web/pages/signup.tsx [hunk 2]
```diff
@@ -124,7 +125,17 @@ export default function Signup({ prepopulateFormValues, token }: inferSSRProps<t
...
+                      required: true,
+                      minLength: {
+                        message: t("password_hint_min"),
+                        value: 7,
+                      },
+                      pattern: {
+                        message: "Should contain a number, uppercase and lowercase letters",
+                        value: /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).*$/gm,
+                      },
+                    })}
```
[the remark this answers is not in the document]
Comment (piyushgarg-dev) [reply apps/web/pages/signup.tsx:138]: Hi @sean-brydon, I borrowed this code from the ChangePassword component. Their plain react-hook-forms's built-in validation was used. Check https://github.com/calcom/cal.com/blob/c610fad11ee1060976721970f7c63f05409327b5/apps/web/pages/settings/security/password.tsx#L145

Are you sure to change this to use zod?

Thanks

[the remark this answers is not in the document]
Comment (piyushgarg-dev) [reply apps/web/pages/signup.tsx]: Pushed at https://github.com/calcom/cal.com/pull/9244/commits/b34147b3b3471ed18c7e96e707845244867c0556

File: apps/web/pages/signup.tsx [hunk 3]
```diff
@@ -21,19 +23,29 @@ import PageWrapper from "@components/PageWrapper";
...
+    email: z.string().email(),
+    password: z
+      .string()
+      .min(7, { message: "Password should have a minimum length of 7 characters" })
+      .regex(/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).*$/, {
+        message: "Password should contain a number, uppercase and lowercase letters",
+      })
+      .refine((value) => value !== "", { message: "Password is required" }),
+    apiError: z.string().optional(),
+  });
```
[the remark this answers is not in the document]
Comment (piyushgarg-dev) [reply apps/web/pages/signup.tsx:42]: I am not able to search for `AppPasswordField` anywhere in the code. Did you mean `PasswordField` being used at 3 places?

[the remark this answers is not in the document]
Comment (hariombalhara) [reply apps/web/pages/signup.tsx:42]: I actually meant that if we create AppPasswordField it can be used at 3 places.

**Pull request closed without merging**